opennebula/resource_opennebula_virtual_machine.go

```diff
@@ -1466,6 +1472,54 @@ func resourceOpennebulaVirtualMachineUpdate(d *schema.ResourceData, meta interfa
 		}
 	}
 
+	if d.HasChange("cpu") || d.HasChange("vcpu") || d.HasChange("memory") {
+		timeout := d.Get("timeout").(int)
+
+		vmState, _, _ := vmInfos.State()
+		vmRequireShutdown := vmState != vm.Poweroff && vmState != vm.Undeployed
+		if vmRequireShutdown {
+			if d.Get("poweroff_hard_on_resize").(bool) {
+				err = vmc.PoweroffHard()
+			} else {
+				err = vmc.Poweroff()
+			}
+			if err != nil {
+				return fmt.Errorf(
+					"Poweroff for virtual machine (ID:%d) failed: %s", vmc.ID, err)
+			}
+			_, err = waitForVMState(vmc, timeout, "POWEROFF")
+			if err != nil {
+				return fmt.Errorf(
+					"waiting for virtual machine (ID:%d) to be in state %s: %s", vmc.ID, "POWEROFF", err)
+			}
+		}
+
+		resizeTpl := dyn.NewTemplate()
+		resizeTpl.AddPair("CPU", d.Get("cpu").(float64))
+		resizeTpl.AddPair("VCPU", d.Get("vcpu").(int))
```

Comment:
We need to add some test if all attributes are not set by the user, for instance:
```
vCPU := d.Get("vcpu").(int)
if vCPU > 0 {
        resizeTpl.AddPair("VCPU", d.Get("vcpu").(int))
 }
```
If we don't do that, OpenNebula receive a template with `VCPU=0` which end with an error from OpenNebula